src/org/nfrac/comportex/cells.cljc

```diff
@@ -666,42 +674,46 @@
                                           min-act :learn-threshold
                                           new-syns :new-synapse-count
                                           max-syns :max-synapse-count
-                                          max-segs :max-segments}]
-  (let [lci-vec (vec lci)] ;; for faster sampling
+                                          max-segs :max-segments
+                                          reuse-frac :relearning-similarity-frac}]
+  (let [any-reuse? (and reuse-frac (pos? reuse-frac) (> (count aci) (count lci)))
+        lci-seq (seq lci)] ;; for faster sampling
     (loop [cells (seq lc)
            m (transient {})
            rng rng]
       (if-let [cell-id (first cells)]
-        (if-let [seg-path (get well-matching-paths cell-id)]
-          ;; choose the well matching segment
-          (recur (next cells)
-                 (assoc! m cell-id (syn/seg-update seg-path :learn nil nil))
-                 rng)
-          ;; otherwise - not well matching - check disconnected synapses
-          (let [cell-segs (p/cell-segments sg cell-id)
-                [match-si exc seg] (best-matching-segment cell-segs aci
-                                                          min-act 0.0)
-                new-segment? (not match-si)
-                [seg-idx die-syns] (if match-si
-                                     [match-si nil]
-                                     (new-segment-id cell-segs pcon max-segs
-                                                     max-syns))
-                grow-n (- new-syns exc)
-                [rng* rng] (if (pos? grow-n) (random/split rng) [rng rng])
-                grow-source-ids (segment-new-synapse-source-ids seg lci-vec grow-n rng*)
-                die-source-ids (if new-segment?
-                                 (keys die-syns) ;; remove any existing (replaced)
-                                 (segment-excess-synapse-source-ids seg grow-n
-                                                                    max-syns))
-                seg-path (conj cell-id seg-idx)]
+        (let [[rng* rng] (if any-reuse? (random/split rng) [rng rng])
+              reuse? (when any-reuse? (< (random/rand-double rng*) reuse-frac))]
```

Comment:
Playing this out in my head, this works when `:perm-init` is above `:perm-connected`, but might need to be a bit different when learning is slower. Otherwise different cells will be selected every time for random reuse, and there will be a higher final reuse-frac than requested. (Definitely not an urgent issue, I just wanted to capture this thought.)

I'm guessing you get what I'm saying, but it's probably useful for me to talk through this. Learn "ABCDE". See "XBCDY". 'B' bursts, and a learning cell is selected. The 'C' from 'abC' is predicted because 'B' bursted. So 'C' doesn't burst. Here we're defining what 'abC' should learn. We're randomly choosing to make it pay attention only to the learnable 'xB' cells, and not the active 'aB' cells which caused this 'abC' cell to be predicted. So we get faster learning of sequences by conflating some cells to mean both 'xbC' and 'abC', but we don't want to overdo it. So I'm pointing out that choosing different random reuse cells on every repetition will cause some chaos.

Off the top of my head, solutions could be:

- Always put `:perm-init` above `:perm-connected`
- Don't make it completely random. Use randomness to add potential synapses, but later always reinforce these segments when they're available. Use a lower reuse-frac to achieve the same final `:relearning-similarity-frac`. So this would be part of choosing the learning segment, always favoring segments that are potentially-connected-to-learnable-cells over segments that are actually-connected-to-nonlearnable-cells, randomly injecting the former into the mix.